torchft/process_group_test.py

```diff
@@ -122,6 +137,122 @@ def check_tensors(arg: Any) -> None:  # pyre-ignore[2]
     return works
 
 
+def _test_multi_pg(pg: ProcessGroup, rank: int, tensor: torch.Tensor) -> None:
+    """
+    Helper function to test a set of collective operations in settings with multiple
+    process groups.
+    """
+    # Test allgather
+    tensor_list = [
+        torch.zeros(2, dtype=torch.int64, device=tensor.device) for _ in range(2)
+    ]
+    ag_tensor = torch.arange(2, dtype=torch.int64, device=tensor.device) + 1 + 2 * rank
+    allgather_work = pg.allgather([tensor_list], [ag_tensor], AllgatherOptions())
+    allgather_work.wait()
+    torch.testing.assert_close(
+        tensor_list[0], torch.tensor([1, 2], device=tensor.device)
+    )
+    torch.testing.assert_close(
+        tensor_list[1], torch.tensor([3, 4], device=tensor.device)
+    )
+
+    # Test allreduce
+    tc = tensor.clone()
+    allreduce_work = pg.allreduce([tc], ReduceOp.SUM)
+    allreduce_work.wait()
+    expected_tensor = torch.tensor([3], device=tc.device)
+    torch.testing.assert_close(tc, expected_tensor)
+
+    # Test allreduce_coalesced
+    tensors = [tensor.clone(), tensor.clone() + 1]
+    allreduce_coalesced_work = pg.allreduce_coalesced(
+        tensors, AllreduceCoalescedOptions()
+    )
+    allreduce_coalesced_work.wait()
+    expected_tensors = [
+        torch.tensor([3], device=tensor.device),
+        torch.tensor([5], device=tensor.device),
+    ]
+    for t, expected in zip(tensors, expected_tensors):
+        torch.testing.assert_close(t, expected)
+
+    # Test all-to-all
+    input_tensor = torch.tensor([rank + 1, rank + 5], device=tensor.device)
+    output_tensor = torch.empty_like(input_tensor)
+    alltoall_work = pg.alltoall_base(
+        output_tensor, input_tensor, [1, 1], [1, 1], AllToAllOptions()
+    )
+    alltoall_work.wait()
+    if rank == 0:
+        expected_alltoall = torch.tensor([1, 2], device=tensor.device)
+    else:
+        expected_alltoall = torch.tensor([5, 6], device=tensor.device)
+    torch.testing.assert_close(output_tensor, expected_alltoall)
+
+    # Test broadcast
+    broadcast_tensor = tensor.clone() if rank == 0 else torch.zeros_like(tensor)
+    broadcast_work = pg.broadcast([broadcast_tensor], BroadcastOptions())
+    broadcast_work.wait()
+    expected_broadcast = torch.tensor([1], device=tensor.device)
+    torch.testing.assert_close(broadcast_tensor, expected_broadcast)
+
+    # Test broadcast_one
+    broadcast_one_tensor = tensor.clone() if rank == 0 else torch.zeros_like(tensor)
+    broadcast_one_work = pg.broadcast_one(broadcast_one_tensor, 0)
+    broadcast_one_work.wait()
+    torch.testing.assert_close(
+        broadcast_one_tensor, torch.tensor([1], device=tensor.device)
+    )
+
+    # Test barrier
+    barrier_work = pg.barrier(BarrierOptions())
+    barrier_work.wait()
+
+    # Test send/recv
+    if rank == 0:
+        send_tensor = tensor.clone()
+        send_work = pg.send([send_tensor], 1, 0)
```

Comment:
Could we also test the actual fault tolerance, i.e. sender fails, sender succeeds but receiver fails, and verify that an appropriate exception / timeout is returned back to the user process? 

And ideally, this is then retried with success after the PG gets reconfigured.